adapters/sonobi/sonobi.go

```diff
@@ -0,0 +1,300 @@
+		sonobiReq.Imp[i].TagID = params.TagID
+	}
+
+	reqJSON, err := json.Marshal(sonobiReq)
+	if err != nil {
+		return nil, err
+	}
+
+	debug := &pbs.BidderDebug{
+		RequestURI: a.URI,
+	}
+	httpReq, _ := http.NewRequest("POST", a.URI, bytes.NewReader(reqJSON))
+	httpReq.Header.Set("Content-Type", "application/json")
+	if sReq.Device != nil {
+		addHeaderIfNonEmpty(httpReq.Header, "User-Agent", sReq.Device.UA)
+		addHeaderIfNonEmpty(httpReq.Header, "X-Forwarded-For", sReq.Device.IP)
+		addHeaderIfNonEmpty(httpReq.Header, "Accept-Language", sReq.Device.Language)
+		addHeaderIfNonEmpty(httpReq.Header, "DNT", strconv.Itoa(int(sReq.Device.DNT)))
+	}
+	sResp, err := ctxhttp.Do(ctx, a.http.Client, httpReq)
+	if err != nil {
+		return nil, err
+	}
+	defer sResp.Body.Close()
+	debug.StatusCode = sResp.StatusCode
+	if sResp.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	body, err := ioutil.ReadAll(sResp.Body)
+	if err != nil {
+		return nil, err
+	}
+	responseBody := string(body)
+	if sResp.StatusCode == http.StatusBadRequest {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("HTTP status %d; body: %s", sResp.StatusCode, responseBody),
+		}
+	}
+
+	if sResp.StatusCode != http.StatusOK {
+		return nil, &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("HTTP status %d; body: %s", sResp.StatusCode, responseBody),
+		}
+	}
+	if req.IsDebug {
+		debug.RequestBody = string(reqJSON)
+		bidder.Debug = append(bidder.Debug, debug)
+		debug.ResponseBody = responseBody
+	}
+
+	var bidResp openrtb.BidResponse
+	err = json.Unmarshal(body, &bidResp)
+	if err != nil {
+		return nil, &errortypes.BadServerResponse{
+			Message: err.Error(),
+		}
+	}
+	bids := make(pbs.PBSBidSlice, 0)
+	for _, sb := range bidResp.SeatBid {
+		for _, bid := range sb.Bid {
+			bidID := bidder.LookupBidID(bid.ImpID)
+			if bidID == "" {
+				return nil, &errortypes.BadServerResponse{
+					Message: fmt.Sprintf("Unknown ad unit code '%s'", bid.ImpID),
+				}
+			}
+
+			adm, _ := url.QueryUnescape(bid.AdM)
+			pbid := pbs.PBSBid{
+				BidID:       bidID,
+				AdUnitCode:  bid.ImpID,
+				BidderCode:  bidder.BidderCode,
+				Price:       bid.Price,
+				Adm:         adm,
+				Creative_id: bid.CrID,
+				Width:       bid.W,
+				Height:      bid.H,
+				DealId:      bid.DealID,
+				NURL:        bid.NURL,
+			}
+			bids = append(bids, &pbid)
+		}
+	}
+
+	sort.Sort(bids)
+	return bids, nil
+}
+
+// MakeRequests Makes the OpenRTB request payload
+func (a *SonobiAdapter) MakeRequests(request *openrtb.BidRequest) ([]*adapters.RequestData, []error) {
+	var errs []error
+	var sonobiExt openrtb_ext.ExtImpSonobi
+	var err error
+
+	var adapterRequests []*adapters.RequestData
+
+	// Sonobi currently only supports 1 imp per request to sonobi.
+	// Loop over the imps from the initial bid request to form many adapter requests to sonobi with only 1 imp.
+	for _, imp := range request.Imp {
+		// Make a copy as we don't want to change the original request
+		reqCopy := *request
```

Comment:
`request` and each `imp` are already shallow copies when passed to the adapters, so no need to recopy them. Good work in trying to avoid race conditions though, thanks! :)

Reply:
I make a copy of the request here so that I don't change the request.Imp that I'm currently iterating over. 

Changing reqCopy to just modifying "request" directly causes only 1 imp to be returned in a bid request.

Reply:
Yes thats the purpose. Each iteration of the incoming request's imps needs to produce multiple outgoing requests to Sonobi that are similar, except with only 1 imp. 

Not copying the incoming request on each iteration causes an issue where the final returned bids back to /openrtb2/auction only include 1 of the potentially many bids.

Reply:
Oh, sorry, I misunderstood your previous comment to mean you were about to change from using reqCopy to modifying request.